### DISCUSSION TITLE
Create custom merge function to support custom utility classes (extended theme)

### DISCUSSION THREAD

**Discussion by charkour** [dcastil/tailwind-merge#48] [Help] [answered]
Hello, thanks for this awesome package.

I noticed an issue with the `twMerge` function when using custom box shadow classes.

```tsx
// Actual
twMerge('shadow-md shadow-lg shadow-light shadow-dark') // 'shadow-lg shadow-light shadow-dark'

// Expected 
twMerge('shadow-md shadow-lg shadow-light shadow-dark') // 'shadow-dark'
```

This works properly with the normal box shadow classes. 

Perhaps I to make a custom tailwind merge function with the light and dark suffixes? Thanks.

**Comment by charkour** [comment]
Inside the `tailwind.config.js` file:

```js
module.exports = {
    mode: 'jit',
    theme: {
        extend: {
            boxShadow: {
                'dark': '0 3px 8px #ccc',
                'light': '0px 2px 5px 0px HSLA(0, 0%, 84.3%, 0.5);',
            },
        },
    },
};

```

**Comment by charkour** [comment]
Looking through the source code, it looks like I need to create a custom tailwind merge function. How would that be done to only add the box shadow custom classes?

**Comment by dcastil** [comment, marked as answer]
Hi @charkour! Indeed, if you're using a custom Tailwind config, you need to use `createTailwindMerge` and create a config for it. You don't need `createTailwindMerge` in all cases, but in this one yes.

Here is how to create it in `v0.7.1`:

```ts
import { createTailwindMerge } from 'tailwind-merge'

export const twMerge = createTailwindMerge((getConfig) => {
    const config = getConfig()

    config.classGroups.shadow.push({ shadow: ['light', 'dark'] })

    return config
})
```

~~In `v0.8.0` which will be released soon you can do it a little cleaner~~ (use https://github.com/dcastil/tailwind-merge/discussions/48#discussioncomment-1516712 instead):

```ts
import { createTailwindMerge, getDefaultConfig, mergeConfigs } from 'tailwind-merge'

export const twMerge = createTailwindMerge(() =>
    mergeConfigs(getDefaultConfig(), {
        classGroups: {
            shadow: [{ shadow: ['light', 'dark'] }],
        },
    })
)
```

If you're wondering where that class group name `shadow` comes from, it is specified in the default config [here](https://github.com/dcastil/tailwind-merge/blob/main/src/default-config.ts). So far there is no straight forward way to figure out which property to access but to look into the source code.

Let me know if that helps.

**Reply by charkour** [reply]
Your version `v0.7.1` almost worked, but calling `push` on a readonly type is restricted.

Here is the solution I came up. 

```tsx
export const customTwMerge = createTailwindMerge((getConfig) => {
    const config = getConfig();
    return {
        ...config,
        classGroups: {
            ...config.classGroups,
            shadow: [{ shadow: [...config.classGroups.shadow[0].shadow, 'light', 'dark'] }],
        },
    };
});
```

**Reply by charkour** [reply]
Initially, I was confused if config was referring to the tailwind config but I realized it is a config for `createTailwindMerge`.

**Reply by charkour** [reply]
Here's an updated version to pull from the `tailwind.config`

```tsx
import { createTailwindMerge } from 'tailwind-merge';
import tailwindConfig from 'tailwind.config';

const customTwMerge = createTailwindMerge((getConfig) => {
    const config = getConfig();
    const customConfig = {
        ...config,
        classGroups: {
            ...config.classGroups,
            shadow: [
                {
                    shadow: [
                        ...config.classGroups.shadow[0].shadow,
                        ...Object.keys(tailwindConfig.theme.extend.boxShadow),
                    ],
                },
            ],
        },
    };
    console.log(customConfig);
    return customConfig;
});
```

I'd like to have a method there I can import the `tailwindConfig` and have all the extended theme classes be applied.

**Reply by dcastil** [reply]
> Initially, I was confused if config was referring to the tailwind config but I realized it is a config for `createTailwindMerge`.

Good to know, I need to make that clearer in the docs.

> I'd like to have a method there I can import the `tailwindConfig` and have all the extended theme classes be applied.

That would be ideal, but the problem is that the tailwind config is meant to run at buildtime and can get really big and [complex](https://tailwindcss.com/docs/theme#referencing-other-values) and can also depend on Node.js modules. [Here](https://github.com/tailwindlabs/tailwindcss/blob/v2.2.17/stubs/defaultConfig.stub.js) you can check out how big the default config of Tailwind really is.

Tailwind also writes about it that the Tailwind config can get big and it's probably not a good idea to put that in client-side code.

> Note that this will transitively pull in a lot of our build-time dependencies, resulting in bigger bundle client-side size. To avoid this, we recommend using a tool like [babel-plugin-preval](https://github.com/kentcdodds/babel-plugin-preval) to generate a static version of your configuration at build-time.

([Source](https://tailwindcss.com/docs/configuration#referencing-in-java-script))

Currently, the only solution is to write a config twice (once for buildtime and once for runtime), while the buildtime config is optimized for convenience and the runtime config is optimized for bundle size. I'm working on making extending the default config easier in #47 and also trying to make the need to create an own tailwind-merge config smaller. But there are still many question marks in that area.

**Reply by dcastil** [reply]
> Your version `v0.7.1` almost worked

Ah yeah, sorry about that, completely forgot about TypeScript there. 😁

**Reply by dcastil** [reply]
I just released v0.8.0. You can now extend your config much simpler:

```ts
import { extendTailwindMerge } from 'tailwind-merge'

const twMerge = extendTailwindMerge({
    classGroups: {
        shadow: [{ shadow: ['light', 'dark'] }]
    }
})
```

**Reply by charkour** [reply]
> Tailwind also writes about it that the Tailwind config can get big and it's probably not a good idea to put that in client-side code.

Yup, I totally agree. That makes sense. I did not think that through initially when I wrote my first comment. 

> Currently, the only solution is to write a config twice (once for buildtime and once for runtime), while the buildtime config is optimized for convenience and the runtime config is optimized for bundle size.

That completely makes sense. I have a somewhat crazy idea, but what if there was a way during build time to fix the tailwind utility class precedence static component classNames. For example, let's say there is a StyledButton and SmallRedButton that extends the BaseButton component. 

```tsx
const BaseButton = ({className}) => {
  return <button className={twMerge("text-red", className)} />
}

const StyledButton = () => {
  return <BaseButton className="text-blue bg-white" />
}

const SmallRedButton = () => {
  return <SmallRedButton className="bg-white w-8" />
}
```

At build time, we know that `<StyledButton />` will have the className of `text-blue bg-white` and `<SmallRedButton />` will have `text-red bg-white w-8`. Therefore, at build time, we could rewrite the components to not be dependent on `<BaseButton />` since the potential for dynamic behavior is actually known at build time.  

```tsx
const StyledButton = () => {
  return <button className="text-blue bg-white" />
}

const SmallRedButton = () => {
  return <button className="text-red bg-white w-8" />
}
```

Writing this out, this is for sure out of scope for this project, and this idea might not even work due to React's rendering method with the VDOM (I'm not sure how Vue's rendering method works). I've just noticed that in many cases at build time the seemingly dynamic output can be known ahead of runtime. What are your thoughts on the subject?

**Reply by charkour** [reply]
> I just released v0.8.0

This works great. Thank you!

**Reply by dcastil** [reply]
I like your idea! I actually see my library as the MVP of a more React-specific version of Tailwind. As you said as well, I think that static analysis is out of scope for this project because that can get hella complicated! 😄 But I hope that someone comes around and builds something that makes React and Tailwind work better together.

**Comment by charkour** [comment]
This is perfect, I had something similar to your `v0.8.0` snippet and then realized `getDefaultConfig` wasn't exported in the latest released version (`v0.7.1`). 

Thanks for the quick response!

**Comment by dcastil** [comment]
Yeah 😄 I also think that this is still way too cumbersome and created https://github.com/dcastil/tailwind-merge/issues/47. That will be in `v0.8.0` as well.

**Comment by dcastil** [comment]
Moving this to discussions as this isn't a bug in tailwind-merge.

**Comment by nenostra** [comment]
Hello @dcastil , if I have custom `space` in my tailwind.config, is there a way to extend twMerge without having to individually define padding, margin, width, height, maxHeight, gap, inset, space, and translate? Same for other tailwind.config values like `borderRadius`. Really appreciate the work you are doing and let me know if I can help in any way to implement something like this.

**Reply by dcastil** [reply]
Hey @nenostra! 👋

At the moment there isn't a way to extend the spacing scale easily unfortunately. As you said, you need to extend all these class groups yourself.

That isn't great, I think this should be easier. I created an issue in https://github.com/dcastil/tailwind-merge/issues/54 to keep track of this. So far I'm not sure how to do this to allow as many scales as possible without increasing the bundle size a lot. I'll think about it in the coming days so I can come up with something decent. If you have any ideas on the topic, feel free to discuss it in the issue!

Would you mind sharing your Tailwind config? Would be interesting to know which spacing and border radius values you use.

**Reply by nenostra** [reply]
Here it is @dcastil, and tyvm for taking a look at this. Will also start reading the codebase and think about how to make something like this easier if possible.
on `theme.js`:
```
const breakpoints = {
  sm: '576px',
  md: '768px',
  lg: '992px',
  xl: '1200px',
};

const spacing = {
  auto: 'auto',
  '0': '0px',
  '2xs': '4px',
  xs: '10px',
  '2sm': '15px',
  sm: '20px',
  md: '24px',
  lg: '30px',
  xl: '40px',
  '2xl': '60px',
  '3xl': '100px',
  '4xl': '130px',
  '5xl': '160px',
  '6xl': '250px',
};

const buttonWidths = {
  'button-default': '185px',
  'button-small': '150px',
  'button-large': '240px',
  'button-fluid': '100%',
};

const containerWidths = {
  'container-md': '750px',
  'container-lg': '970px',
  'container-xl': '1170px',
};

const modalWidths = {
  'large-modal-lg': '920px',
  'large-modal-md': '730px',
  'default-modal': '730px',
  'small-modal': '540px',
};

const heights = {
  header: '60px',
};

const palette = {
  black: '#000000',
  'light-navy': '#fafbfc',
  navy: '#325e77',
  'dark-navy': '#0f2532',
  'lighter-green': '#c2f7e4',
  'light-green': '#5bbe9b',
  green: '#3ee4a9',
  'dark-green': '#3aa680',
  'light-gray': '#fafbfc',
  gray: '#d6e1e4',
  'dark-gray': '#8ea0a4',
  red: '#ff5661',
  'dark-red': '#c71d28',
  'green-translucent': 'rgba(62, 228, 169, 0.8)',
  'dark-navy-translucent': 'rgba(15, 37, 50, 0.9)',
  orange: '#ff9c00',
  white: '#fff',
  transparent: 'transparent',
};

const fontWeight = {
  regular: 400,
  semibold: 600,
};

const letterSpacing = {
  lg: '0.5px',
  sm: '0.3px',
};

const fontSize = {
  'h1-lg': '54px',
  'h1-md': '42px',
  'h1-sm': '32px',

  'h2-lg': '36px',
  'h2-md': '28px',
  'h2-sm': '26px',

  'h3-lg': '24px',
  'h3-md': '22px',
  'h3-sm': '20px',

  body: '16px',
  p1: '12px',
  p2: '10px',
};

const lineHeight = {
  'h1-lg': '62px',
  'h1-md': '50px',
  'h1-sm': '38px',

  'h2-lg': '44px',
  'h2-md': '32px',
  'h2-sm': '32px',

  'h3-lg': '32px',
  'h3-md': '30px',
  'h3-sm': '28px',

  body: '24px',
  p1: '20px',
  p2: '14px',
};

const boxShadow = {
  DEFAULT: '0px 1px 6px 0px rgba(50, 94, 119, 0.26)',
  card: '0 2px 30px 0 rgba(0, 0, 0, 0.4)',
  button: '0 5px 10px rgba(62, 228, 169, 0.3)',
};

const borderRadius = {
  none: '0',
  small: '4px',
  DEFAULT: '8px',
  full: '9999px',
};

const transitionDuration = { hover: '400ms' };
```

then on `tailwind.config`:
```
theme: {
    screens: breakpoints,
    fontFamily: {
      sans: ['Montserrat', 'sans-serif'],
    },
    colors: palette,
    fontSize,
    spacing,
    lineHeight,
    fontWeight,
    letterSpacing,
    borderRadius,
    transitionDuration,
    extend: {
      height: {
        ...heights,
      },
      width: {
        ...buttonWidths,
        ...containerWidths,
        ...modalWidths,
      },
    },
  },
```
